aten/src/ATen/native/ReplicationPadding.cpp

```diff
@@ -211,12 +210,621 @@ Tensor& replication_pad1d_backward_out_cpu_template(
+{
+  int iStartX = std::max(0, -pad_l);
+  int iStartY = std::max(0, -pad_t);
+  int oStartX = std::max(0, pad_l);
+  int oStartY = std::max(0, pad_t);
+
+  int64_t k, ip_x, ip_y;
+#pragma omp parallel for private(k, ip_x, ip_y)
+  for (k = 0; k < nslices; k++)
+  {
+    for (int64_t i = 0; i < oheight; i++) {
+      for (int64_t j = 0; j < owidth; j++) {
+        if (j < pad_l) {
+          ip_x = pad_l;
+        } else if (j >= pad_l && j < iwidth + pad_l) {
+          ip_x = j;
+        } else {
+          ip_x = iwidth + pad_l - 1;
+        }
+        ip_x = ip_x - oStartX + iStartX;
+
+        if (i < pad_t) {
+          ip_y = pad_t;
+        } else if (i >= pad_t && i < iheight + pad_t) {
+          ip_y = i;
+        } else {
+          ip_y = iheight + pad_t - 1;
+        }
+        ip_y = ip_y - oStartY + iStartY;
+
+        scalar_t *dest_p = output_p + k*owidth*oheight + i * owidth + j;
+        scalar_t *src_p = input_p + k*iwidth*iheight + ip_y * iwidth + ip_x;
+        *dest_p = *src_p;
+      }
+    }
+  }
+}
+
+void replication_pad2d_out_cpu_template(Tensor& output,
+    const Tensor& input,
+    IntList paddingSize)
+{
+  int pad_l = paddingSize[0];
+  int pad_r = paddingSize[1];
+  int pad_t = paddingSize[2];
+  int pad_b = paddingSize[3];
+  int dimw = 2;
+  int dimh = 1;
+  int dimslices = 0;
+  int64_t nbatch = 1;
+  int64_t nslices;
+  int64_t iheight;
+  int64_t iwidth;
+  int64_t oheight;
+  int64_t owidth;
+
+  AT_CHECK(input.numel() > 0 && (input.dim() == 3 || input.dim() == 4),
+      "3D or 4D (batch mode) tensor expected for input, but got: ", input);
+
+  if (input.dim() == 4)
+  {
+    nbatch = input.size(0);
+    dimw++;
+    dimh++;
+    dimslices++;
+  }
+
+  /* sizes */
+  nslices = input.size(dimslices);
+  iheight = input.size(dimh);
+  iwidth = input.size(dimw);
+  oheight = iheight + pad_t + pad_b;
+  owidth  = iwidth + pad_l + pad_r;
+
+  AT_CHECK(owidth >= 1 || oheight >= 1,
+      "input (H: ", iheight, ", W: ", iwidth, " )is too small."
+      " Calculated output H: ", oheight, " W: ", owidth);
+
+
+  /* get contiguous input */
+  auto input_ = input.contiguous();
+
+  /* resize output */
+  if (input_.dim() == 3)
+  {
+    output.resize_({nslices, oheight, owidth});
+    AT_DISPATCH_FLOATING_TYPES(input_.type(), "replication_pad2d", [&] {
+        auto input_data = input_.data<scalar_t>();
+        auto output_data = output.data<scalar_t>();
+        replication_pad2d_out_frame<scalar_t> (input_data, output_data,
             nslices,
-            iwidth,
-            owidth,
-            pad_l, pad_r);
+            iwidth, iheight,
+            owidth, oheight,
+            pad_l, pad_r,
+            pad_t, pad_b);
+        }
+        );
+  }
+  else
+  {
+    int64_t p;
+
+    output.resize_({nbatch, nslices, oheight, owidth});
+
+#pragma omp parallel for private(p)
+    for (p = 0; p < nbatch; p++)
+    {
+      AT_DISPATCH_FLOATING_TYPES(input_.type(), "replication_pad2d", [&] {
```

Comment:
As I said in the first PR (https://github.com/pytorch/pytorch/pull/15507#discussion_r243773621) you should do the `AT_DISPATCH` macro before you OMP parallelize. The code should look something like:

```
AT_DISPATCH_FLOATING_TYPES(input_.type(), "replication_pad2d", [&] {
  auto input_data = input_.data<scalar_t>();
  auto output_data = output.data<scalar_t>();
#pragma omp parallel for private(p)
  for (p = 0; p < nbatch; p++) {
    ...
  }
}
```

This didn't need to be explicitly handled in the old code, because everything in the file was "pre-dispatched" (e.g., `scalar_t` is statically known.) That's effectively the same as `AT_DISPATCH_FLOATING_TYPES` having lived at the very top-level of the entire file, not pushed into the loop as it is here.